[RFC] Remarkable port

tcrs opened the pull request:
I have a remarkable (https://remarkable.com/). The builtin PDF reader was too slow so I did a quick port of koreader. This is mostly based on the Sony PRS port.

I've never worked on koreader before so would appreciate any comments etc.

I think at the moment the main thing that is missing is the wifi manager stuff. wpa_supplicant is set up as a dbus service which xochitl (the interface the remarkable ships with) talks to via dbus. The wpa_cli control socket seems to be disabled.

I'll open a pull request in koreader-base too.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/5697)
<!-- Reviewable:end -->

pazos commented:
I miss the platform code. I mean, at least, the script that launchs koreader. I guess you're running using a launcher and calling `/reader.lua /some/path` yourself. 

The script that launchs KO is the one that does OTA updates, export tessdata and font paths and a few things more, so it might be interesting for your device too.

I used the Sony PRS port too as a reference to build for Cervantes, but it is lacking some key features: (how to install it?, how to run it?) that can make easier for others to enjoy the new supported device.

NiLuJe commented on frontend/device/remarkable/event_map.lua:
```diff
@@ -0,0 +1,7 @@
+    [106] = "RPgFwd",
+    [116] = "Power",
+}
+-- TODO libremarkable has 143 as "wakeup" - don't know what that corresponds to?
```
Does it happen to have a hall effect sensor? In which case, my guess would be the key event that's sent by that when the magnet is lifted off ;).

tcrs replied:
Ah possibly, I'll try running a magnet around the device and see if I get any events. I don't have the remarkable case so I'd not thought about that possibility.

tcrs replied:
I did try running a magnet around the bezels but I couldn't get any events out of it

tcrs commented on kodev:
```diff
@@ -117,6 +117,7 @@ SUPPORTED_TARGETS="
     kobo
     cervantes
     sony-prstux
+	remarkable
```
I've fixed up (hopefully) all the hard tabs

tcrs commented on frontend/ui/uimanager.lua:
```diff
@@ -1155,7 +1186,7 @@ end
 -- suspension.
 function UIManager:suspend()
-    if Device:isCervantes() or Device:isKobo() or Device:isSDL() or Device:isSonyPRSTUX() then
+    if Device:isCervantes() or Device:isKobo() or Device:isSDL() or Device:isSonyPRSTUX() or Device:isRemarkable() then
```
I've re-ordered the remarkable code to before the sony code everywhere it seemed sensible to do so to me.

tcrs commented:
> I miss the platform code. I mean, at least, the script that launchs koreader. I guess you're running using a launcher and calling `/reader.lua /some/path` yourself.
> 
> The script that launchs KO is the one that does OTA updates, export tessdata and font paths and a few things more, so it might be interesting for your device too.
> 
> I used the Sony PRS port too as a reference to build for Cervantes, but it is lacking some key features: (how to install it?, how to run it?) that can make easier for others to enjoy the new supported device.

I'm currently launching koreader from a systemd unit (which sets a couple of env vars and runs `luajit reader.lua`) and using a trivially patched version of https://github.com/dixonary/button-capture-reMarkable to start it. I have the koreader systemd service set up to conflict with the xochitl one so systemd automatically stops xochitl when koreader starts, and the koreader service starts the xochitl service when koreader exits.

I'll take a look at the other platforms and see about sketching something out for the remarkable port.

tcrs commented:
I am getting there with the platform integration stuff. I'd like to test OTA updates work correctly. Has anyone got any tips on how to test that? Can I set up a server on my PC to serve an "update" to koreader running on my remarkable?

Frenzie commented:
You can, all you need is the zsync files.

You can see the command here: https://github.com/koreader/koreader-misc/blob/68a1fa175b94cd807c440fbdf91db878f2a913a0/nightswatcher/nightswatcher.py#L219-L220

The servers are hardcoded here:
https://github.com/koreader/koreader/blob/f60b4c36ce00abfe0c80d5d2146f38562dd69eef/frontend/ui/otamanager.lua#L22-L35

But you could also just switch the platform to Kobo for testing. It'd overwrite your files with Kobo stuff but since you're using the same TC it should still run. Just without updating your platform-specific files. ;-)

tcrs commented:
Thanks for the info regarding the updates. I've got it mostly working, but zsync keeps segfaulting on the remarkable. I get this log:
```
./spinning_zsync: line 44:  2327 Segmentation fault      (core dumped) ./zsync "$@"
zsync -o ./ota/koreader.updated.tar -u http://192.168.1.2:80/ ./ota/koreader-remarkable-latest-stable.zsync
gzip: unrecognized option '--no-name'
BusyBox v1.30.1 (2019-10-15 13:27:35 UTC) multi-call binary.

Usage: gzip [-cfkdt] [FILE]...
./spinning_zsync: line 44:  2588 Segmentation fault      (core dumped) ./zsync "$@"
```
As you can see I added an echo to the spinning_zsync script to print out the zsync command just before it runs it.

I'm guessing the problem is that the gzip on the remarkable is from busybox which doesn't have the `--no-name` argument. I added GNU gzip as a thirdparty thing and bundled it (like tar is currently), which seems to work fine. Does that seem sensible?

Frenzie commented:
Neither does Kobo busybox, so I wouldn't be inclined to think of that as a likely cause. Besides which, isn't zlib statically compiled into zsync? Pinging @NiLuJe

tcrs commented:
Hmm OK, I don't know the details but I can see in `zsync/c/libzsync/zsync.c` the function `zsync_recompress` shells out to `gzip`. It does look like the arguments which are passed to the `gzip` invocation depend on the zsync file header, so maybe it has something to do with the way I'm making the zsync file for the update. I'm running:
```
zsyncmake koreader-remarkable-v2019.11-92-g46a35526_2020-01-23.targz -u 'http://192.168.1.2:80/koreader-remarkable-v2019.11-92-g46a35526_2020-01-23.targz' -o koreader-remarkable-latest-stable.zsync
```
(I just copied the command from nightswatcher.py)

Frenzie commented:
The regular one starts like this:
```
zsync: 0.6.2
Filename: koreader-kobo-v2019.12.targz
MTime: Mon, 30 Dec 2019 16:42:27 +0000
Blocksize: 2048
Length: 73943040
Hash-Lengths: 2,2,5
Z-URL: koreader-kobo-v2019.12.targz
SHA-1: f60bc1065ffbb69963dfc71199278ff7ca104580
Z-Map2: 45271
```

What does yours say? (You could also attach the whole thing I suppose.)

tcrs commented:
Ah ha, mine is quite different:
```
zsync: 0.6.2
Safe: Z-Filename Recompress MTime
Z-Filename: koreader-remarkable-v2019.11-96-g2d0c9608_2020-01-25.targz
Filename: koreader-remarkable-v2019.11-96-g2d0c9608_2020-01-25.targz
MTime: Sat, 25 Jan 2020 16:20:36 +0000
Blocksize: 2048
Length: 72796160
Hash-Lengths: 2,2,5
Z-URL: http://192.168.1.2:80/koreader-remarkable-v2019.11-96-g2d0c9608_2020-01-25.targz
SHA-1: c2dfdc05af93d9ac33ab4df95b4befed383b4a05
Recompress: 1f8b0800000000000003 --no-name
Z-Map2: 54013
```

If I use the `-Z` option to `zsyncmake` (which apparently stops it treating gzipped files specially) I get
```
zsync: 0.6.2
Filename: koreader-remarkable-v2019.11-96-g2d0c9608_2020-01-25.targz
MTime: Sat, 25 Jan 2020 16:20:36 +0000
Blocksize: 2048
Length: 35962128
Hash-Lengths: 2,2,5
URL: http://192.168.1.2:80/koreader-remarkable-v2019.11-96-g2d0c9608_2020-01-25.targz
SHA-1: af9ba6016b2d3e8141dc266cdaafb527eca3879c
```
If I use `-C` instead I get
```
zsync: 0.6.2
Filename: koreader-remarkable-v2019.11-96-g2d0c9608_2020-01-25.targz
MTime: Sat, 25 Jan 2020 16:20:36 +0000
Blocksize: 2048
Length: 72796160
Hash-Lengths: 2,2,5
Z-URL: http://192.168.1.2:80/koreader-remarkable-v2019.11-96-g2d0c9608_2020-01-25.targz
SHA-1: c2dfdc05af93d9ac33ab4df95b4befed383b4a05
Z-Map2: 54013
```
Which seems to be pretty much the same as the regular one. I wonder why, it looks like the zsync version is the same and the nightswatcher script doesn't seem to pass any extra options. I'll check that the OTA update works without GNU gzip with these zsync files, but I suspect they will.

Not sure what to make of that.

Frenzie commented:
Yeah, it has to be that `Recompress` there. Maybe it has something to do with the wider system having a file association for .targz the same as .tar.gz? The nightswatcher Docker image runs on Ubuntu 16.04 iirc, unless I already updated it to 18.04.

NiLuJe commented:
Note that I'm currently investigating moving to zsync2 to avoid a bunch of weird corner-cases issue with the now abandoned zsync ;).

NiLuJe commented:
But to answer the question: yeah, there's a weird trick to how the zsync files are generated re compression or not, and I have no actual clue how that's supposed to work, I just now it does :D.

(i.e., the zsync files points to a gzipped archive, but zsync generates an uncompressed tarball. It's weird.).

tcrs commented:
Well the OTA updates seem to be working with busybox gzip now if I use the `-C` option to `zsyncmake`. I'm running Arch Linux, FWIW I have version `0.6.2-4` of `zsync`.

pazos commented on frontend/device/remarkable/device.lua:
```diff
@@ -0,0 +1,125 @@
+    return ffi.string("reMarkable")
+end
+
+return Remarkable:new{
```
You can move elements to L9 and simply
```lua
return Remarkable
``` 
here.

tcrs replied:
OK,done

pazos commented on frontend/device/remarkable/device.lua:
```diff
@@ -0,0 +1,125 @@
+    os.execute("systemctl reboot")
+end
+
+function Remarkable:getSoftwareVersion()
```
is this used?

tcrs replied:
Nope, it's a leftover from copy-pasting the sony-prstux code. Removed

pazos commented on frontend/device/remarkable/device.lua:
```diff
@@ -0,0 +1,125 @@
+    return ffi.string("ZeroGravitas")
+end
+
+function Remarkable:getDeviceModel()
```
is this used?

tcrs replied:
Nope, it's a leftover from copy-pasting the sony-prstux code. Removed

pazos commented on frontend/device/remarkable/powerd.lua:
```diff
@@ -0,0 +1,32 @@
+
+local Remarkable_PowerD = BasePowerD:new{
+    is_charging = nil,
+    fl_min = 0,
```
fl_min and fl_max can be removed on devices without hasFrontlight

tcrs replied:
OK, removed

NiLuJe commented on frontend/device/remarkable/device.lua:
```diff
@@ -0,0 +1,116 @@
+    isTouchDevice = yes,
+    hasKeys = yes,
+    hasFrontlight = no,
+    display_dpi = 166,
```
```suggestion
    display_dpi = 212,
```

NiLuJe commented on frontend/device/remarkable/device.lua:
```diff
@@ -0,0 +1,116 @@
+        local Screensaver = require("ui/screensaver")
+        Screensaver:close()
+        local UIManager = require("ui/uimanager")
+        UIManager:nextTick(function() UIManager:setDirty("all", "full") end)
```
Is it actually needed? Or does it double-flash on resume?

I can't recall the full code flow, but we got rid of one of those on Kobo because it was already taken care of by the ScreenSavers widget.

NiLuJe replied:
Ah, see  `needsScreenRefreshAfterResume` in `Device` ;).

NiLuJe commented on frontend/device/remarkable/powerd.lua:
```diff
@@ -0,0 +1,30 @@
+local BasePowerD = require("device/generic/powerd")
+
+-- TODO older firmware doesn't have the -0 on the end of the file path
```
Probably won't be able to run the binaries because of an older glibc (I presume) anyway ;).

NiLuJe commented on platform/remarkable/koreader.sh:
```diff
@@ -0,0 +1,233 @@
+        # The idea for the margins being to leave enough room for an fbink -Z bar, small horizontal margins, and a font size based on what 6pt looked like @ 265dpi
+        ./fbink -q -b -O -t regular=./fonts/droid/DroidSansMono.ttf,top=$((viewHeight/2 + FONTH * 2 + FONTH/2)),left=$((viewWidth/60)),right=$((viewWidth/60)),px=$((viewHeight/64)) "${crashLog}"
+        # So far, we hadn't triggered an actual screen refresh, do that now, to make sure everything is bundled in a single flashing refresh.
+        ./fbink -q -f -s top=0,left=0
```
```suggestion
        ./fbink -q -f -s
```

(No behavior change, that was just a quirk needed w/ older FBInk versions because `getopt` is weird).

NiLuJe commented:
LGTM, too, besides the very minor nitpicks above ;).

(base has been bumped since w/ reMarkable support, hence the conflict).

pazos commented:
@tcrs: Shall I fix conflicts and merge this or wait for you?.  Not in a hurry but I want to fix issues on other platforms introduced when we merged the base part of this PR without waiting for the frontend part to be ready.

Frenzie commented:
@pazos Unfortunately it looks like you'll have to copy the branch into a new PR due to a lack of push access. I just tried committing @NiLuJe's suggestions.

pazos commented:
@Frenzie: yeah, I did that to fix conflicts in https://github.com/pazos/koreader/tree/remarkable (keeping the rest of commits as they were). Just need to commit these two suggestions and it is ready to merge.

Frenzie commented:
Please merge when ready :+1:

pazos commented:
I will wait a few hours for @tcrs feedback and merge before next nightly otherwise
